juju/utils.py

```diff
@@ -181,6 +181,10 @@ async def run_with_interrupt(task, *events, log=None):
                                         return_when=jasyncio.FIRST_COMPLETED)
     for f in pending:
         f.cancel()  # cancel unfinished tasks
+        try:
+            await f
```

Comment:
Shouldn't this be a loop around cancel, and *then* a loop around await? That way everything has time to start shutting down before we wait on them? Otherwise if the first is slow, the second never gets started, etc.

Reply:
Yeah that's a better way of doing it. I'm trying to understand why the test keeps getting stuck (only fails on CI, locally it seems to be working). I'm not sure, but it might have something to do with this, actually, I'll see if fixing this will change the test behavior https://github.com/juju/python-libjuju/actions/runs/6029145138/job/16358032056?pr=939